### PR TITLE
AWS#267 Added support for Port Annotation in ServiceExport Object

### PR DESCRIPTION
**Pull request by sesmaili**
**What type of PR is this?** Feature

**Which issue does this PR fix**: AWS#267

**What does this PR do / Why do we need it**: This PR enables users to define which port/container to be exported vi ServiceExport object.

**If an issue # is not available please add repro steps and logs from aws-gateway-controller showing the issue**:

**Testing done on this change**: Yes
make presubmit output:
```go
?       github.com/aws/aws-application-networking-k8s/pkg/aws   [no test files]
?       github.com/aws/aws-application-networking-k8s/pkg/deploy        [no test files]
ok      github.com/aws/aws-application-networking-k8s/pkg/aws/services  0.017s  coverage: 2.9% of statements
ok      github.com/aws/aws-application-networking-k8s/pkg/config        0.011s  coverage: 77.4% of statements
?       github.com/aws/aws-application-networking-k8s/pkg/k8s   [no test files]
?       github.com/aws/aws-application-networking-k8s/pkg/model/lattice [no test files]
?       github.com/aws/aws-application-networking-k8s/pkg/runtime       [no test files]
?       github.com/aws/aws-application-networking-k8s/pkg/utils [no test files]
?       github.com/aws/aws-application-networking-k8s/pkg/utils/retry   [no test files]
?       github.com/aws/aws-application-networking-k8s/pkg/utils/ttime   [no test files]
ok      github.com/aws/aws-application-networking-k8s/pkg/deploy/lattice        0.042s  coverage: 86.1% of statements
**ok      github.com/aws/aws-application-networking-k8s/pkg/gateway       0.054s  coverage: 77.8% of statements**
ok      github.com/aws/aws-application-networking-k8s/pkg/latticestore  0.004s  coverage: 73.3% of statements
ok      github.com/aws/aws-application-networking-k8s/pkg/model/core    0.004s  coverage: 40.9% of statements
ok      github.com/aws/aws-application-networking-k8s/pkg/model/core/graph      0.003s  coverage: 17.2% of statements
ok      github.com/aws/aws-application-networking-k8s/pkg/utils/log     0.002s  coverage: 0.0% of statements [no tests to run]
```
E2E Tests output:
```go
Ran 7 of 7 Specs in 3104.031 seconds
SUCCESS! -- 7 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestIntegration (3104.19s)
PASS
ok      github.com/aws/aws-application-networking-k8s/test/suites/integration   3104.210s
```

**Automation added to e2e**: No

**Will this PR introduce any new dependencies?**: No

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**: Yes

**Does this PR introduce any user-facing change?**: Yes.

```release-note
Users will be able to define which ports/containers to expose through ServiceExport by adding an annotation to ServiceExport objects.
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by zijun726911** [comment]
Seems after you change, for example for this connectivity set up:

```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: inventory
spec:
  parentRefs:
  - name: my-hotel
    sectionName: http
  rules:
  - backendRefs:
    - name: inventory-ver1
      kind: Service
      port: 8090
    matches:
    - path:
        type: PathPrefix
        value: /ver1
  - backendRefs:
    - name: inventory-ver2
      kind: ServiceImport
      port: 9090
    matches:
    - path:
        type: PathPrefix
        value: /ver2
```
 
The behaviour would be:
for inventory-ver1, if will regardless of HTTPRoute backendRefs‘s inventory-ver1 's port  as before. And use inventory-ver1  k8sService's (all) port(s) blindly

for inventory-ver2, if will regardless of HTTPRoute backendRefs‘s inventory-ver2 's port as before And use the intersection ports for inventory-ver2's k8sservice's (all) port(s)  and inventory-ver2's serviceExport's annotation (currently support on port only).


That a good workaround to narrow down vpcLattice target's ports to be created for serviceExport. But for the long tern should we regard to use HTTPRoute backendRefs‘ ports to create lattice targets? (whatever backendRef kind service or backendRef kind ServiceImport)

**Comment by liwenwu-amazon** [comment]
This PR fixes #267

File: pkg/gateway/model_build_targets.go [hunk 1]
```diff
@@ -119,15 +138,19 @@ func (t *latticeTargetsModelBuildTask) buildLatticeTargets(ctx context.Context)
...
 				for _, port := range endPoint.Ports {
 					glog.V(6).Infof("serviceReconcile-endpoints: address %v, port %v\n", address, port)
 					target := latticemodel.Target{
 						TargetIP: address.IP,
 						Port:     int64(port.Port),
 					}
-					targetList = append(targetList, target)
+					if portAnnotations == 0 || int64(target.Port) == portAnnotations {
+						targetList = append(targetList, target)
+						fmt.Printf("portAnnotations:%v, target.Port:%v\n", portAnnotations, target.Port)
```
Comment (liwenwu-amazon) [inline pkg/gateway/model_build_targets.go:150]: can u use glog.V(6).Infof here instead of fmt.Printf()

File: pkg/gateway/model_build_targets.go [hunk 2]
```diff
@@ -107,7 +109,24 @@ func (t *latticeTargetsModelBuildTask) buildLatticeTargets(ctx context.Context)
 		errmsg := fmt.Sprintf("Build Targets failed because K8S service %v does not exist", namespacedName)
 		return errors.New(errmsg)
 	}
+
+	portAnnotations := int64(0)
```
Comment (liwenwu-amazon) [inline pkg/gateway/model_build_targets.go:113]: NIT,  use a constant (e.g. #define NO_PORT_SELECTOR,  and portAnnotations := NO_PORT_SELECTOR) to make it more readable

**Review by liwenwu-amazon — COMMENTED** [review]
LGTM, and left some minor comments

**Comment by liwenwu-amazon** [comment]
> Seems after your change, for example for this connectivity set up:
> 
> ```
> apiVersion: gateway.networking.k8s.io/v1beta1
> kind: HTTPRoute
> metadata:
>   name: inventory
> spec:
>   parentRefs:
>   - name: my-hotel
>     sectionName: http
>   rules:
>   - backendRefs:
>     - name: inventory-ver1
>       kind: Service
>       port: 8090
>     matches:
>     - path:
>         type: PathPrefix
>         value: /ver1
>   - backendRefs:
>     - name: inventory-ver2
>       kind: ServiceImport
>       port: 9090
>     matches:
>     - path:
>         type: PathPrefix
>         value: /ver2
> ```
> 
> The behaviour would be: for inventory-ver1, if will regardless of HTTPRoute backendRefs‘s inventory-ver1 's port (as before). And use inventory-ver1 k8sService's (all) port(s) blindly (as before)
> 
> for inventory-ver2, if will regardless of HTTPRoute backendRefs‘s inventory-ver2 's port (as before). And use the intersection ports for inventory-ver2's k8sservice's (all) port(s) and inventory-ver2's serviceExport's annotation (currently support one port only).
> 
> That a good workaround to narrow down vpcLattice target's ports to be created for serviceExport. But for the long tern should we regard to use HTTPRoute backendRefs‘ ports to create lattice targets? (whatever backendRef kind service or backendRef kind ServiceImport)

IMO, for long team, we should push this port specification into `ServiceExport` object definition.

File: pkg/gateway/model_build_targets.go (hunk 1 above)
Comment (sesmaili) [reply pkg/gateway/model_build_targets.go:150]: Done

File: pkg/gateway/model_build_targets.go (hunk 2 above)
Comment (sesmaili) [reply pkg/gateway/model_build_targets.go:113]: Done

**Comment by sesmaili** [comment]
> This PR fixes #267 . Fixes still needed for #241

Are you sure? Looks more like #241 as I read the bug.

**Review by liwenwu-amazon — APPROVED** [review]
Thank you!

**Pull request merged**